### ISSUE TITLE
Uncaught Error: error sending request for url

### ISSUE THREAD

**Issue by knightmarehs** [a16z-infra/ai-town#249]
The agents can't generate conversations, the traceback is as follows:

8/12/2024, 5:33:59 PM [CONVEX A(aiTown/agentOperations:agentGenerateMessage)] Uncaught Error: error sending request for url (http://127.0.0.1:11434/api/embeddings): error trying to connect: tcp connect error: Connection refused (os error 61)

**Comment by ianmacartney** [comment]
check that you have something running on
http://127.0.0.1:11434/api/embeddings - e.g. ollama - and that it's
accessible from where you're running the backend

**Comment by uannyao** [comment]
Hi there, I'm facing a similar issue, the agents can't generate conversations, but in the traceback, it's saying the request was forbidden:

9/29/2024, 2:53:34 PM [CONVEX A(aiTown/agentOperations:agentGenerateMessage)] Uncaught Error: Request to http://localhost:11434/api/embeddings forbidden

**Comment by likeUMR** [comment]
I have the same problem too!!!

**Comment by quanchentg** [comment]
@likeUMR did you resolve this issue or figure something out? Thanks.

**Comment by quanchentg** [comment]
@ianmacartney  is there any clue for this issue? Thanks.

**Comment by ianmacartney** [comment]
Sorry I haven't seen this issue before. I suggest dropping in logs,
reproducing simple requests, verifying some assumptions (is ai-town running
on your machine or in the cloud environment? is it running in a container?
is ollama running somewhere accessible?
You might also try 127.0.0.1 instead of `localhost` since some Windows
users have seen odd behavior there. Best of luck with the debugging

**Comment by leol15** [comment]
For me this happened when the frontend is pointing to Convex backend managed by Convex in the cloud (dashboard.convex.dev).

Check your `.env.local` file and make sure `VITE_CONVEX_URL=http://127.0.0.1:3210` is there. If not you can re-generate it with `npm run dev:backend` after clearing the contents of `.env.local`.

Then you might need to restart the frontend app - `npm run dev:frontend` to pick up the change.

**Comment by quanchentg** [comment]
@leol15 Thanks for your nice suggestions. Let me give it a try!

**Comment by Tsailj** [comment]
@leol15 您好，想问您是使用本地的convex后端，不使用云端的convex就能成功跟Ollama通信是吗，我现在遇到的问题就是跟Ollama的连接被拒绝，我现在的convex使用的是云端的dashboard.convex.dev，.env.local也是设置的云端的没有用http://127.0.0.1:3210，您是使用http://127.0.0.1:3210就能避免通信被拒这个问题是吗

**Comment by Tsailj** [comment]
> Uncaught Error: Request to http://127.0.0.1:11434/api/embeddings forbidden
> 
> When i change the embeddings to embed(base on a post about the ollama new api doc, so it is http://127.0.0.1:11434/api/embed) the problem still exist!

Hello, I’m facing the same issue as you. I also changed ‘embedding’ to ‘embed’, but I’m still getting an error（Uncaught Error: Request to http://localhost:11434/api/embed forbidden
）. Have you found a solution? Thank you.

**Comment by leol15** [comment]
@Tsailj Yes, local backend worked for me. This is the relevant command to start the backend https://github.com/a16z-infra/ai-town/blob/main/Justfile#L31

 I'm unsure how the cloud backend was able to communicate with Ollama running locally, appreciate if others could explain!

**Comment by ianmacartney** [comment]
in the readme there are directions for using ngrok / tunnelmole to proxy
the traffic locally

**Comment by chuanxin** [comment]
> @leol15 您好，想问您是使用本地的convex后端，不使用云端的convex就能成功跟Ollama通信是吗，我现在遇到的问题就是跟Ollama的连接被拒绝，我现在的convex使用的是云端的dashboard.convex.dev，.env.local也是设置的云端的没有用[http://127.0.0.1:3210，您是使用http://127.0.0.1:3210就能避免通信被拒这个问题是吗](http://127.0.0.1:3210%EF%BC%8C%E6%82%A8%E6%98%AF%E4%BD%BF%E7%94%A8http://127.0.0.1:3210%E5%B0%B1%E8%83%BD%E9%81%BF%E5%85%8D%E9%80%9A%E4%BF%A1%E8%A2%AB%E6%8B%92%E8%BF%99%E4%B8%AA%E9%97%AE%E9%A2%98%E6%98%AF%E5%90%97)

我只能成功的使用雲端的 convex，再另外設置 tunnel 服務連到本地端的 localhost:11434，tunnel 的連結是寫入到 convex dashboard 的 Settings 中，新增一項 OLLAMA_HOST 環境變數，以建立 ngrok 連結為例，需要指定發送服務的標頭內 host 內容為 localhost:11434，指令參考: `ngrok http 11434 --host-header="localhost:11434"`，我還另外在本地端環境設置了 `Ollama_ORIGINS` 變數為 `*`

**Comment by ImLaoBJie** [comment]
> > [@leol15](https://github.com/leol15) 您好，想问您是使用本地的convex后端，不使用云端的convex就能成功跟Ollama通信是吗，我现在遇到的问题就是跟Ollama的连接被拒绝，我现在的convex使用的是云端的dashboard.convex.dev，.env.local也是设置的云端的没有用[http://127.0.0.1:3210，您是使用http://127.0.0.1:3210就能避免通信被拒这个问题是吗](http://127.0.0.1:3210%EF%BC%8C%E6%82%A8%E6%98%AF%E4%BD%BF%E7%94%A8http://127.0.0.1:3210%E5%B0%B1%E8%83%BD%E9%81%BF%E5%85%8D%E9%80%9A%E4%BF%A1%E8%A2%AB%E6%8B%92%E8%BF%99%E4%B8%AA%E9%97%AE%E9%A2%98%E6%98%AF%E5%90%97)
> 
> 我只能成功的使用雲端的 convex，再另外設置 tunnel 服務連到本地端的 localhost:11434，tunnel 的連結是寫入到 convex dashboard 的 Settings 中，新增一項 OLLAMA_HOST 環境變數，以建立 ngrok 連結為例，需要指定發送服務的標頭內 host 內容為 localhost:11434，指令參考: `ngrok http 11434 --host-header="localhost:11434"`，我還另外在本地端環境設置了 `Ollama_ORIGINS` 變數為 `*`

Thanks, this method is effective. I dont know why convex cannot assign the ollama request on the local. I dont use the recommended ngrok or Tunnelmole, they r so slow with less quota. I have an Oray Huashengke account, so i use it to generate a tunnel from my windows Ollama port `127.0.0.1:11434`, and set `just convex env set OLLAMA_HOST <tunnel address>` on wsl, it works.
